scripts/thumbnail-generator/generate_thumbnail.py

```diff
@@ -0,0 +1,302 @@
+#!/usr/bin/env python3
+
+from pxr import Usd, UsdGeom, UsdMedia, Sdf, Gf, UsdUtils, UsdLux, UsdShade
+import subprocess
+import math
+import os
+import sys
+import argparse
+from pathlib import Path
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="This script takes a thumbnail image of the given USD file supplied and associates it with the file.")
+    parser.add_argument('usd_file', 
+                        type=str, 
+                        help='The USD file you want to add a thumbnail to. If USDZ is input, a new USD file will be created to wrap the existing one called <input>_Thumbnail.usd')
+    parser.add_argument('--dome-light',
+                        type=str,
+                        help='The path to the dome light HDR image to use, if any')
+    parser.add_argument('--create-usdz-result', 
+                        action='store_true',
+                        help='Returns the resulting files as a new usdz file called <input>_Thumbnail.usdz')
+    parser.add_argument('--width',
+                        type=int,
+                        help='The width of the image to generate. Default is 2048.',
+                        default=2048)
+    parser.add_argument('--height',
+                        type=int,
+                        help='The height of the image to generate. Default is 2048. If height is not specified, the image is square.')
+    parser.add_argument('--output-extension', 
+                        type=str, 
+                        help='The file extension of the output image you want (exr, png..). If using exr, make sure your usd install includes OpenEXR',
+                        default='png')
+    parser.add_argument('--verbose', 
+                        action='store_true',
+                        help='Prints out the steps as they happen')
+    return parser.parse_args()
+
+def generate_thumbnail(usd_file, verbose, extension):
+    if verbose: 
+        print("Step 1: Setting up the camera...")
+    
+    subject_stage = Usd.Stage.Open(usd_file)
+    subject_file = usd_file
+
+    if (UsdGeom.GetStageUpAxis(subject_stage) == 'Z'):
+        subject_stage = generate_y_up_stage(subject_stage, usd_file)
+        subject_file = 'y_up.usda'
+
+    setup_camera(subject_stage, subject_file)
+    
+    if verbose:
+        print("Step 2: Taking the snapshot...")
+    
+    Path("renders").mkdir(parents=True, exist_ok=True)
+
+    image_path = os.path.join("renders", create_image_filename(usd_file, extension)).replace("\\", "/")
+    image_name = take_snapshot(image_path)
+
+    return image_name
+
+def generate_y_up_stage(stage, usd_file):
```

Comment:
Both the need for and the implementation of this method seem highly suspect to me... Can't you just rotate the camera and lights differently depending on whether the asset is y-up or z-up? That seems like it would be a lot easier than trying to re-orient the asset.

If you did want to make a y-up copy of the stage (and I still think that you don't), you could sublayer the root layer of stage onto the root layer of the new stage, and then you'd just need to author the rotations on the root prims of the new layer. You wouldn't need to worry about re-binding all the materials.

Reply:
I originally had this as a camera rotation, but then it did require a light rotation and then I wasn't sure what other things I might be missing on needing to rotate, so applying one parent object rotation seemed easier, until Blender exported Materials at the top level and it required this more cumbersome method. 

Thanks for a definitive opinion on it because both seemed in theory to be about the same level of work, I'll add those rotations back in for the camera and lights and make sure everything works the same